core.network/src/org/netbeans/core/network/utils/IpAddressUtils.java

```diff
@@ -0,0 +1,501 @@
+
+            if (segments.length != 4) {
+                return false;
+            }
+            
+            for (String segment : segments) {
+                if (segment == null || segment.length() == 0) {
+                    return false;
+                }
+                
+                // leading zeroes are not allowed within the segment.
+                if (segment.length() > 1 && segment.startsWith("0")) {
+                    return false;
+                }
+                
+                try {
+                    int value = Integer.parseInt(segment);
+                    if (value > 255) {
+                        return false;
+                    }
+                } catch (NumberFormatException e) {
+                    return false;  // Unlikely. Already matched by regexp
+                }
+            }
+            return true;
+        }
+        return false;
+    }
+
+    /**
+     * Does a shallow check if the argument looks like an IPv6 address as
+     * opposed to a host name. Note, that a return value of {@code true} doesn't
+     * guarantee that the argument is a <i>valid</i> IPv6 literal, but a return
+     * value of {@code false} is a guarantee that it is not.
+     * 
+     * @param ipAddressStr
+     * @return true if argument looks like an IPv6 literal.
+     */
+    public static boolean looksLikeIpv6Literal(String ipAddressStr) {
+        if (ipAddressStr == null) {
+            return false;
+        }
+        // ::d is the shortest possible form of an IPv6 address.
+        if (ipAddressStr.length() < 3) {
+            return false;
+        }
+        if (ipAddressStr.startsWith(":") || ipAddressStr.endsWith(":")) {
+            return true;
+        }
+        // Matches A80::8:800:200C:417A or 0A80::8:800:200C:417A
+        if (ipAddressStr.length() >= 5) {
+            if ((ipAddressStr.charAt(3) == ':') || (ipAddressStr.charAt(4) == ':')) {
+                return true;
+            }
+        }
+        return false;
+    }
+    
+    /**
+     * Does a shallow check if the argument looks like an IPv4 address (on the
+     * form {@code #.#.#.#}) as opposed to a host name. Note, that a return
+     * value of {@code true} doesn't guarantee that the argument is a
+     * <i>valid</i> IPv4 literal, but a return value of {@code false} is a 
+     * guarantee that it is not.
+     *
+     * @param ipAddressStr
+     * @return true if argument looks like an IPv4 literal
+     *     
+     */
+    public static boolean looksLikeIpv4Literal(String ipAddressStr) {
+        if (ipAddressStr == null || ipAddressStr.isEmpty()) {
+            return false;
+        }
+        if (!(isAsciiDigit(ipAddressStr.charAt(0)))) {
+            return false;
+        }
+        int dotPos = ipAddressStr.indexOf('.');
+        if (dotPos > 0) {
+            for(int i = 0; i < dotPos; i++) {
+                if (!(isAsciiDigit(ipAddressStr.charAt(0)))) {
```

Comment:
The index should be _i_ not _0_. In general: is this really faster, than `IPV4_PATTERN.matcher(ipAddressStr).matches()`?

Reply:
The bug fixed. Thx.

Re use of regexp. The IPV4_PATTERN is used to check for full validity. And it requires a full IPv4 pattern, e.g. `#.#.#.#`  as opposed to say `#.#` which in Java world is also a valid IPv4 address. 

The two `looksLike`  methods have a very narrow use case. I've tried to explain that in the Javadoc (improved). They are not validity tests. If the user actually wants full validity check he should use `isValidIpv4Address()` as opposed to `looksLikeIpv4Literal()`. The former is actually now unused in the module but I've kept it for completeness and because it may have value for those implementing their own version of  PAC evaluator. This is also the reason why the two `looksLike` methods are not private, even if they have a very narrow use case.

Finding out if a string is a host name or an IP literal is somewhat difficult. The problem is that an all digit string is actually a valid hostname (what wise guy came up with that RFC??), but a digit is not allowed as the first character in a domain name. Example: "123.45" is a valid IPv4 address, but it is not a valid hostname, because of the '4'. However, the corner case value of "123" is indeterminate. This is explained in the Javadoc.